### PR TITLE
Issue 226 - Added an option for getting analysis result in JSON format

### PR DESCRIPTION
**Pull request by seena-mathew**
1. Created a listener for JSONReporter.
2. Added an option for selecting the type of analysis report

### CODE REVIEW

**Comment by kruegers** [comment]
Hey @enriozuni and @rakshitkr,

Please review the code and test the feature on your machines. Thanks.

**Comment by enriozuni** [comment]
> @seena-mathew While reviewing your PR changes I noticed that most files in your commits are shown as entirely modified. But when I used online tool like diffchecker I found out that changes are made only to certain sections of original file. Please take care of it. Only the changes that you've made should be reflected in these commits.

Yeah, I noticed that too.

File: CryptoAnalysis/src/main/java/crypto/reporting/JSONReporter.java [hunk 1]
```diff
@@ -113,16 +117,15 @@ public void afterAnalysis() {
...
 	
-		if (outputFolder != null) {
 		try {
 			 ObjectMapper mapper = new ObjectMapper();
 			 ObjectWriter writer = mapper.writer(new DefaultPrettyPrinter());
 			 writer.writeValue(Paths.get(outputFolder + File.separator+"CogniCrypt-JSON-Report.json").toFile(), json);
+			 LOG.info("JSON Report generated to file : "+ outputFolder + File.separator+"CogniCrypt-JSON-Report.json");
 	     } 
 		catch (IOException e) {
 	    	 e.printStackTrace();
```
Comment (rakshitkr) [inline CryptoAnalysis/src/main/java/crypto/reporting/JSONReporter.java:127]: Use LOG.error here instead

File: CryptoAnalysis/src/main/java/crypto/reporting/CommandLineReporter.java [hunk 2]
```diff
@@ -1,112 +1,119 @@
...
+import soot.util.EscapedWriter;
+
+public class CommandLineReporter extends ErrorMarkerListener {
+	private static final Logger LOG = LoggerFactory.getLogger(CommandLineReporter.class);
+	private Format reportformat;
+	private File outputFolder;
+	private List<CrySLRule> rules;
+	private Collection<IAnalysisSeed> objects = new HashSet<>();
+
+	public CommandLineReporter(String string, Format format, List<CrySLRule> rules) {
```
Comment (rakshitkr) [inline CryptoAnalysis/src/main/java/crypto/reporting/CommandLineReporter.java:36]: Rename first argument in this constructor

**Comment by rakshitkr** [comment]
@kruegers I've test the below scenarios
* Without giving reportDir and reportFormat -> must display the result in CLI :: `SUCCESS`
* With reportDir and without reportFormat -> txt report must be generated in the specified reportDir :: `SUCCESS`
* Without reportDir and with reportFormat -> report with specified format must be generated in directory of CryptoAnalysis-2.7.2-SNAPSHOT-jar-with-dependencies.jar :: `SUCCESS`
* With reportDir and reportFormat -> report with specified format must be generated in specified directory :: `SUCCESS`
* Provide the format of the report --reportFormat (txt, json, sarif) :: `SUCCESS`

**Comment by enriozuni** [comment]
@seena-mathew I just finished testing your PR on my machine. Almost everything is working as it is supposed to, but I have some small comments and advice on the .txt file below on what you could update. Afterward, I can check your code one more time :-) 

[testing-report.txt](https://github.com/CROSSINGTUD/CryptoAnalysis/files/4628954/testing-report.txt)

File: CryptoAnalysis/src/main/java/crypto/HeadlessCryptoScanner.java [hunk 3]
```diff
@@ -1,464 +1,494 @@
...
+				callGraphAlogrithm = CG.CHA;
+			}
+		} else {
+			callGraphAlogrithm = CG.CHA;
+		}
+		
+		final Format reportFormat;
+		if (options.hasOption("reportFormat")) {
+			String format =  options.getOptionValue("reportFormat");
+			if(format.equalsIgnoreCase("csv")) {
```
[the remark this answers is not in the document]
Comment (seena-mathew) [reply CryptoAnalysis/src/main/java/crypto/HeadlessCryptoScanner.java]: Changed if-else statement to switch case

File: CryptoAnalysis/src/main/java/crypto/reporting/JSONReporter.java (hunk 1 above)
Comment (seena-mathew) [reply CryptoAnalysis/src/main/java/crypto/reporting/JSONReporter.java:127]: It is not an error. It is just a message to display the report location in CLI.

File: CryptoAnalysis/src/main/java/crypto/reporting/CommandLineReporter.java (hunk 2 above)
Comment (seena-mathew) [reply CryptoAnalysis/src/main/java/crypto/reporting/CommandLineReporter.java]: Renamed the first argument of the constructor in CommandLineReporter and SARIFReporter

**Comment by seena-mathew** [comment]
> @seena-mathew I just finished testing your PR on my machine. Almost everything is working as it is supposed to, but I have some small comments and advice on the .txt file below on what you could update. Afterward, I can check your code one more time :-)
> 
> [testing-report.txt](https://github.com/CROSSINGTUD/CryptoAnalysis/files/4628954/testing-report.txt)

@enriozuni For the TXT report format, along with the text report, a Jimple file is generated and the result also gets printed in CLI. For the SARIF report format, the generated report is also a text file with the contents in SARIF format. Both the above functionalities were as it is before incorporating the JSON report.

**Comment by johspaeth** [comment]
Hi all,
I do not understand the intention of this pull request. Is there any reason for not using / extending the SARIF exporter? [SARIF](https://docs.oasis-open.org/sarif/sarif/v2.0/sarif-v2.0.html) is a static analysis exchange format and CryptoAnalysis can already output it as `.json`. I would rather stick to one format.

File: CryptoAnalysis/src/main/java/crypto/reporting/JSONReporter.java (hunk 1 above)
Comment (rakshitkr) [reply CryptoAnalysis/src/main/java/crypto/reporting/JSONReporter.java:127]: Sorry.. The suggested change was for `e.printStackTrace();`

**Comment by seena-mathew** [comment]
@johspaeth Issue 226 was created to 'Add possibility to output results into a standard format'. We were not aware that the SARIF can be extended to JSON format. That's why another option was added for JSON format.

**Pull request closed without merging**